apps/heft/src/operations/runners/PhaseOperationRunner.ts

```diff
@@ -0,0 +1,97 @@
+// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
+// See LICENSE in the project root for license information.
+
+import { performance } from 'perf_hooks';
+import { AlreadyReportedError } from '@rushstack/node-core-library';
+
+import { OperationStatus } from '../OperationStatus';
+import { deleteFilesAsync, type IDeleteOperation } from '../../plugins/DeleteFilesPlugin';
+import type { IOperationRunner, IOperationRunnerContext } from '../IOperationRunner';
+import type { HeftPhase } from '../../pluginFramework/HeftPhase';
+import type { HeftPhaseSession } from '../../pluginFramework/HeftPhaseSession';
+import type { HeftTaskSession, IHeftTaskCleanHookOptions } from '../../pluginFramework/HeftTaskSession';
+import type { InternalHeftSession } from '../../pluginFramework/InternalHeftSession';
+
+export interface IPhaseOperationRunnerOptions {
+  internalHeftSession: InternalHeftSession;
+  phase: HeftPhase;
+}
+
+export class PhaseOperationRunner implements IOperationRunner {
+  public readonly silent: boolean = true;
+
+  private readonly _options: IPhaseOperationRunnerOptions;
+
+  public get name(): string {
+    return `Phase "${this._options.phase.phaseName}"`;
+  }
+
+  public constructor(options: IPhaseOperationRunnerOptions) {
+    this._options = options;
+  }
+
+  public async executeAsync(context: IOperationRunnerContext): Promise<OperationStatus> {
+    const { internalHeftSession, phase } = this._options;
+    const { clean, cleanCache, watch } = internalHeftSession.parameterManager.defaultParameters;
+
+    // Load and apply the plugins for this phase only
+    const phaseSession: HeftPhaseSession = internalHeftSession.getSessionForPhase(phase);
+    const { phaseLogger, cleanLogger, cleanHook } = phaseSession;
+    phaseLogger.terminal.writeVerboseLine('Applying task plugins');
+    await phaseSession.applyPluginsAsync();
+
+    // Avoid running the phase operation when in watch mode
+    if (watch) {
+      return OperationStatus.Success;
+    }
+
+    // Run the clean hook
+    if (clean) {
```

Comment:
Can this be refactored into something on a base class? This looks very similar to the lifecycle version.

Reply:
Tried, doesn't work all that gracefully to be honest. They're similar, but implementation details differ enough that it doesn't make much sense.